launch/test/launch/utilities/test_type_utils.py

```diff
@@ -0,0 +1,578 @@
+    assert is_instance_of([1, 2], List[int], True)
+    assert is_instance_of(['1', 2], List[int], True)
+    assert not is_instance_of(['1', 2.], List[int], True)
+    assert not is_instance_of([1, 2.], List[int], True)
+    assert not is_instance_of([1, 2], int, True)
+    assert not is_instance_of([1, '2'], List[str], True)
+
+    with pytest.raises(ValueError):
+        is_instance_of(1, bytes)
+    with pytest.raises(ValueError):
+        is_instance_of([1], List)
+    with pytest.raises(ValueError):
+        is_instance_of([True, False], list)
+    with pytest.raises(ValueError):
+        is_instance_of([True, False], Union[int, str])
+
+
+@pytest.mark.parametrize(
+    'coerce_to_type_impl',
+    (
+        coerce_to_type,
+        lambda value, data_type=None, can_be_str=False: get_typed_value(
+            value, data_type, can_be_str),
+    ),
+    ids=[
+        'testing coerce_to_type implementation',
+        'testing get_typed_value implementation',
+    ]
+)
+def test_coercions_using_yaml_rules(coerce_to_type_impl):
+    assert coerce_to_type_impl('asd') == 'asd'
+    assert coerce_to_type_impl('tRuE') == 'tRuE'
+    assert coerce_to_type_impl("'1'") == '1'
+    assert coerce_to_type_impl("'off'") == 'off'
+
+    assert coerce_to_type_impl('1') == 1
+    assert coerce_to_type_impl('1000') == 1000
+
+    assert coerce_to_type_impl('1.') == 1.0
+    assert coerce_to_type_impl('1000.0') == 1000.
+    assert coerce_to_type_impl('0.2') == .2
+    assert coerce_to_type_impl('.3') == 0.3
+
+    assert coerce_to_type_impl('on') is True
+    assert coerce_to_type_impl('off') is False
+    assert coerce_to_type_impl('True') is True
+
+    assert coerce_to_type_impl('[2, 1, 1]') == [2, 1, 1]
+    assert coerce_to_type_impl('[.2, .1, .1]') == [.2, .1, .1]
+    assert coerce_to_type_impl('[asd, bsd, csd]') == ['asd', 'bsd', 'csd']
+    assert coerce_to_type_impl('[on, false, no]') == [True, False, False]
+
+    assert coerce_to_type_impl('[asd, 2.0]', can_be_str=True) == ['asd', 2.0]
+
+
+@pytest.mark.parametrize(
+    'coerce_to_type_impl',
+    (
+        coerce_to_type,
+        lambda value, data_type=None, can_be_str=False: get_typed_value(
+            value, data_type, can_be_str),
+    ),
+    ids=[
+        'testing coerce_to_type implementation',
+        'testing get_typed_value implementation',
+    ]
+)
+def test_coercions_given_specific_type(coerce_to_type_impl):
+    assert coerce_to_type_impl('asd', data_type=str) == 'asd'
+    assert coerce_to_type_impl('tRuE', data_type=str) == 'tRuE'
+    assert coerce_to_type_impl("'1'", data_type=str) == "'1'"
+    assert coerce_to_type_impl("'off'", data_type=str) == "'off'"
+    assert coerce_to_type_impl("''1''", data_type=str) == "''1''"
+    assert coerce_to_type_impl('{1}', data_type=str) == '{1}'
+
+    assert coerce_to_type_impl('1', data_type=int) == 1
+    assert coerce_to_type_impl('1000', data_type=int) == 1000
+
+    assert coerce_to_type_impl('1', data_type=float) == 1.0
```

Comment:
@ivanpauno hmm, so coercion is flexible enough to accept an integer representation when `float` is explicitly given as the expected data type, but otherwise fails (even when a `List[float]` is specified)? I find it surprising at least.

Reply:
> @ivanpauno hmm, so coercion is flexible enough to accept an integer representation when float is explicitly given as the expected data type, but otherwise fails (even when a List[float] is specified)? I find it surprising at least.


I think that when `List[float]` is explicitly specified we could accept both ints and floats members.
IMO, we can make this more permissive in the future.

Reply:
Yeap. I'd be slightly inclined towards making it consistent (uniform) no, but I won't block on it.